### PR TITLE
Remove Noop programs and record payload requirements

### PR DESCRIPTION
**Pull request by raychu86**
## Motivation

**This PR implements a few key features for testnet3:**
 - The ProgramID CRH curve updated from BW6 to EdwardsBls12
 - Record payloads are now optional
 - Program IDs are now optional in a transition.
     - Executions now do not need a program id, program path, program vk, and program proof for transfers/noop transitions.
     - This vastly reduces the size of transactions.
     - Actual applications outside of transfers will require real program executions will corresponding proofs.
 - Usage/reliance of the noop program has been removed.
    - The Noop program is no longer embedded into the Network trait
    - Noop parameters parameters are no longer regenerated and stored
 - Inner circuit updated to reflect the optional executions and optional record payloads.
    
**Additional updates:** 
 - Parameter have been regenerated
 - Updated constants (record size, record ciphertext size, program id size, etc)
 - Updated serialization size tests

### CODE REVIEW

File: dpc/src/virtual_machine/execution.rs [hunk 1]
```diff
@@ -88,48 +105,60 @@ impl<N: Network> Execution<N> {
...
+                    eprintln!("Failed to validate the program proof: {:?}", error);
                     false
                 }
-            },
-            Err(error) => {
-                eprintln!("Failed to validate the program proof: {:?}", error);
-                false
             }
+        } else {
+            true
```
Comment (howardwu) [inline dpc/src/virtual_machine/execution.rs:127]:
```suggestion
            // TODO (howardwu): CRITICAL - Check the program ID is a noop. (Also rearchitect program ID out of ProgramExecution)
            true
```

**Review by howardwu — APPROVED** [review]

**Comment by howardwu** [comment]
LGTM.

I'm marking this ready and merging into `feat/records` as the changeset is already quite large (and correct!).

There are a few cases, notably with Program ID, which could be simplified, and while Serde does not handle serialization of `Option<T>`, it appears to be warranted for our case now given how frequent this comes up (and normalized of a representation we have).

In any case, we can iterate on the remaining WIPs in a subsequent PR.

File: dpc/src/circuits/inner_circuit.rs [hunk 2]
```diff
@@ -534,11 +537,17 @@ impl<N: Network> ConstraintSynthesizer<N::InnerScalarField> for InnerCircuit<N>
 
                 let given_value = Int64::alloc(&mut declare_cs.ns(|| "given_value"), || Ok(record.value().as_i64()))?;
 
-                let given_payload =
-                    UInt8::alloc_vec(&mut declare_cs.ns(|| "given_payload"), &record.payload().to_bytes_le()?)?;
+                // Use an empty payload if the record does not have one.
+                let payload = if let Some(payload) = record.payload().clone() { payload } else { Payload::default() };
```
Comment (Pratyush) [inline dpc/src/circuits/inner_circuit.rs]:
```suggestion
                let payload = record.payload().unwrap_or(&Payload::default());
```

File: dpc/src/circuits/inner_circuit.rs [hunk 3]
```diff
@@ -534,11 +537,17 @@ impl<N: Network> ConstraintSynthesizer<N::InnerScalarField> for InnerCircuit<N>
...
+                let given_payload = UInt8::alloc_vec(&mut declare_cs.ns(|| "given_payload"), &payload.to_bytes_le()?)?;
 
-                let given_program_id =
-                    UInt8::alloc_vec(&mut declare_cs.ns(|| "given_program_id"), &record.program_id().to_bytes_le()?)?;
+                // Use an empty program id if the record does not have one.
+                let program_id_bytes = if let Some(program_id) = record.program_id() {
+                    program_id.to_bytes_le()?
+                } else {
+                    vec![0u8; N::PROGRAM_ID_SIZE_IN_BYTES]
+                };
```
Comment (Pratyush) [inline dpc/src/circuits/inner_circuit.rs]:
```suggestion
                let program_id_bytes = record.program_id().map_or(Ok(vec![0u8; N::PROGRAM_ID_SIZE_IN_BYTES]), |program_id| program_id.to_bytes_le())?;
```

File: dpc/src/circuits/inner_circuit.rs [hunk 4]
```diff
@@ -676,9 +685,14 @@ impl<N: Network> ConstraintSynthesizer<N::InnerScalarField> for InnerCircuit<N>
 
             // Allocate the program ID.
             let executable_program_id_field_elements = {
+                let program_id_bytes = if let Some(program_id) = public.program_id {
+                    program_id.to_bytes_le()?
+                } else {
+                    vec![0u8; N::PROGRAM_ID_SIZE_IN_BYTES]
+                };
```
Comment (Pratyush) [inline dpc/src/circuits/inner_circuit.rs]: Same here
```suggestion
            let program_id_bytes = record.program_id().map_or(Ok(vec![0u8; N::PROGRAM_ID_SIZE_IN_BYTES]), |program_id| program_id.to_bytes_le())?;
```

File: dpc/src/record/ciphertext.rs [hunk 5]
```diff
@@ -49,11 +49,16 @@ impl<N: Network> Ciphertext<N> {
         randomizer: N::RecordRandomizer,
         record_view_key_commitment: N::RecordViewKeyCommitment,
         record_bytes: Vec<u8>,
-        program_id: N::ProgramID,
+        program_id: Option<N::ProgramID>,
     ) -> Result<Self, RecordError> {
+        let program_id_bytes = match program_id {
```
Comment (Pratyush) [inline dpc/src/record/ciphertext.rs]:
```suggestion
        let program_id_bytes = program_id.map_or(Ok(vec![0u8; N::PROGRAM_ID_SIZE_IN_BYTES]), |program_id| program_id.to_bytes_le())?;
 ```

File: dpc/src/transition/request.rs [hunk 6]
```diff
@@ -183,16 +183,16 @@ impl<N: Network> Request<N> {
         let program_id = {
             let program_id = OnceCell::new();
             for record in &self.records {
-                if record.program_id() != *N::noop_program_id() && program_id.set(record.program_id()).is_err() {
+                if record.program_id() != None && program_id.set(record.program_id()).is_err() {
```
Comment (Pratyush) [inline dpc/src/transition/request.rs]:
```suggestion
                if record.program_id().is_some() && program_id.set(record.program_id()).is_err() {
```

File: dpc/src/transition/request.rs [hunk 7]
```diff
@@ -288,14 +283,14 @@ impl<N: Network> Request<N> {
     }
 
     /// Returns the program ID.
-    pub fn to_program_id(&self) -> Result<N::ProgramID> {
+    pub fn to_program_id(&self) -> Result<Option<N::ProgramID>> {
         let program_id = OnceCell::new();
         for record in &self.records {
-            if record.program_id() != *N::noop_program_id() && program_id.set(record.program_id()).is_err() {
+            if record.program_id() != None && program_id.set(record.program_id()).is_err() {
```
Comment (Pratyush) [inline dpc/src/transition/request.rs]:
```suggestion
            if record.program_id().is_some() && program_id.set(record.program_id()).is_err() {
```

File: dpc/src/virtual_machine/execution.rs [hunk 8]
```diff
@@ -88,48 +105,61 @@ impl<N: Network> Execution<N> {
...
+                &ProgramPublicVariables::new(transition_id),
+                &program_execution.program_proof,
+            ) {
+                Ok(is_valid) => match is_valid {
+                    true => true,
+                    false => {
+                        eprintln!("Program proof failed to verify");
+                        false
+                    }
+                },
```
Comment (Pratyush) [inline dpc/src/virtual_machine/execution.rs]:
```suggestion
                Ok(true) => true,
                Ok(false) => {
                    eprintln!("Program proof failed to verify");
                    false
                },
```

**Pull request merged**